#5731

nad opened the pull request:
Changes related to issue #5731.

nad committed:
[ #5731 ] Added a separate meta-store for solved meta-variables.

And optimised some pieces of code.

nad committed:
Made the code in Agda.Utils.Cluster more general.

nad committed:
[ #5731 ] Added the field metaModule to MetaId.

And changed the type of the field metaId from Nat to Word64.

When a MetaId is converted to JSON the metaModule field is included in
the output.

nad committed:
Removed some dead code.

To avoid having to refactor it.

nad committed:
Made "cabal repl" work.

nad committed:
Made NamesIn more general.

Now the class can be used to find meta-variables as well.

nad committed:
Changed the implementation of traverseTermM for Dummy.

nad committed:
Added some NamesIn instances.

nad committed:
Changed the implementation of namesAndMetasIn'.

For the Defn constructors Function, Record and PrimitiveSort, and the
Term constructor Dummy.

nad committed:
[ #5731 ] Added --save-metas and --no-save-metas.

I received some help from Andreas Abel, Jesper Cockx and Ulf Norell.

nad committed:
Added instantiateWhen.

andreasabel commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -1328,13 +1358,32 @@ data Frozen
+
+-- | Information about remote meta-variables.
+
+data RemoteMetaVariable = RemoteMetaVariable
```
"remote" sound to me like a metavariable accessible via a URI.  (Cf. _remote procedure call_).
How about "external"?

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
+  ss        <- created stSolvedMetaStore nextMeta
+  return (a, LocalMetaStores { openMetas = os, solvedMetas = ss })
+  where
+  created :: Lens' LocalMetaStore TCState -> MetaId -> m LocalMetaStore
```
I'd find it more natural to pass the `LocalMetaStore` to this auxiliary function rather than the lens.  This would mean a duplication of `useTC`, but it would be a more meaningful factor.

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
+    mv@Just{} -> return mv
+    Nothing   -> lkup <$> useR stOpenMetaStore
+  where
+  lkup = MapS.lookup m
```
Please inline `lkup` for sake of readability.

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
+-- | Find information about the (local or remote) meta-variable, if
+-- any.
+
+lookupMeta ::
```
Breaking a long type signature as follows:
```haskell
lookupMeta 
  :: ReadTCState m 
  => MetaId 
  -> m (Maybe (Either RemoteMetaVariable MetaVariable))
```
seems to make more sense as then haddockumentation of arguments can be attached on the right, or after each line.

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
+  ReadTCState m =>
+  MetaId -> m (Maybe (Either RemoteMetaVariable MetaVariable))
+lookupMeta m = do
+  mv <- lookupLocalMeta' m
```
Why not?
```haskell
  lookupLocalMeta' m >>= \case
```
Saves binding `mv` (and shadowing it then).

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
+lookupMetaInstantiation ::
+  ReadTCState m => MetaId -> m MetaInstantiation
+lookupMetaInstantiation m = do
+  mi <- lookupMeta m
```
`>>= \case`

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
 
-getMetaPriority :: (MonadFail m, ReadTCState m) => MetaId -> m MetaPriority
-getMetaPriority = mvPriority <.> lookupMeta
+getMetaPriority ::
```
Document that this only looks in the local metas.

nad replied:
There are more examples of undocumented procedures with changed semantics, but I can fix this one.

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -120,83 +126,198 @@ dontAssignMetas cont = do
 getMetaType m = do
-  mv <- lookupMeta m
-  return $ case mvJudgement mv of
+  j <- lookupMetaJudgement m
```
`lookupMetaJudgement m <&> \case`

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -240,11 +361,10 @@ instance IsInstantiatedMeta a => IsInstantiatedMeta (Abs a) where
-    InstV tel v -> Just $ foldr mkLam v tel
-    _           -> Nothing
-
+  inst <- lookupMetaInstantiation m
```
`lookupMetaInstantiation m <&> \case`

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -588,44 +716,48 @@ clearMetaListeners m =
 
 instance UnFreezeMeta MetaId where
-  unfreezeMeta x = do
+  unfreezeMeta x = unlessM (($ x) <$> isRemoteMeta) $ do
```
`isRemoteMeta <*> pure x`

andreasabel commented on src/full/Agda/TypeChecking/Reduce.hs:
```diff
@@ -91,9 +103,9 @@ simplify = liftReduce . simplify'
-  mv <- lookupMeta m
-  case mvInstantiation mv of
-    InstV _tel v -> noMetas <$> instantiateFull v
+  inst <- lookupMetaInstantiation m
```
`\case`

andreasabel commented on src/full/Agda/TypeChecking/Reduce.hs:
```diff
@@ -139,34 +151,64 @@ instance (Instantiate a, Instantiate b) => Instantiate (a,b) where
+ifPredicateDoesNotHoldFor ::
+  MetaId -> ReduceM a -> ReduceM a -> ReduceM a
+ifPredicateDoesNotHoldFor m doesNotHold holds = do
+  pred <- redPred <$> askR
```
```haskell
caseMaybeM (redPred <$> askR) holds $ \ p ->
  ifM (p m) holds doesNotHold
```

andreasabel reviewed (changes requested):
Nice work!

Could you please put an summary/description of the most salient features of this PR in at the top of the "Conversation" thread?

Also, the distinction between local/external metas should be documented in such a summary somewhere in the codebase, maybe in `TC.Monad.Base` where the meta-store is introduced.

nad commented:
> Could you please put an summary/description of the most salient features of this PR in at the top of the "Conversation" thread?

I put a [summary](https://github.com/agda/agda/issues/5731#issuecomment-1026283941) in the thread of issue #5731.

> Also, the distinction between local/external metas should be documented in such a summary somewhere in the codebase, maybe in `TC.Monad.Base` where the meta-store is introduced.

It is documented:

https://github.com/agda/agda/blob/0e80e2b75490d272724db07c80efe865ae89d6eb/src/full/Agda/TypeChecking/Monad/Base.hs#L1476-L1482

However, perhaps the documentation could be improved:

https://github.com/agda/agda/blob/0e80e2b75490d272724db07c80efe865ae89d6eb/src/full/Agda/TypeChecking/Monad/Base.hs#L985-L986

I'll fix this.

Most of your other feedback is related to the style of the code. Thanks for your effort, but I don't want to spend time on bike-shedding. I asked you to review my changes because I removed some dead code (you have in the past objected to this), and because I changed the `size-solver` program.

andreasabel commented on src/size-solver/test/fail.test:
```diff
@@ -6,7 +6,7 @@ Size constraint solver (C) 2013 Andreas Abel and Felix Reihl
 j' < ∞
 j < X
-Error: flexible variables are not allowed in hypotheses
+Error
```
The error message is lost by the refactoring.  Should be refactored in a way that does not make the error message unusable in the context of `size-solver`.

andreasabel commented on src/full/Agda/TypeChecking/SizedTypes/Syntax.hs:
```diff
@@ -170,14 +175,19 @@ instance Plus (SizeExpr' r f) Offset (SizeExpr' r f) where
 
+-- | Error messages produced by the constraint simplification monad.
+
+type Error = TCM Doc
```
Mmh, this does not seem to work in the context of `size-solver`.

I thought about alternatives back and forth, and the best seems to be a structured `Error` type with one constructor for each error.  Then, printing can be handled by the client, either using `PrettyTCM` or something simpler.

andreasabel reviewed (changes requested):
- Ok to delete the old size-solver.
- The errors of the size solver should be printable in the `size-solver` client.
- Please place a commentary on the new local/remote meta distinction in the code, explaining purpose, design decisions, and potential caveats.

andreasabel commented:
> > Could you please put an summary/description of the most salient features of this PR in at the top of the "Conversation" thread?
> 
> I put a [summary](https://github.com/agda/agda/issues/5731#issuecomment-1026283941) in the thread of issue #5731.

This is very informative, I'd like to see this also in the codebase.

> Most of your other feedback is related to the style of the code. Thanks for your effort, but I don't want to spend time on bike-shedding. I asked you to review my changes because I removed some dead code (you have in the past objected to this), and because I changed the `size-solver` program.

You could have stated which part of the code you want the feedback on.
(See my second review for this feedback.)

nad commented:
> * The errors of the size solver should be printable in the `size-solver` client.

Why do we support this program? I already wasted some time on making it compile, and I do not plan to do anything else.

> * Please place a commentary on the new local/remote meta distinction in the code, explaining purpose, design decisions, and potential caveats.

Does the following sound reasonable (as part of the documentation of `RemoteMetaVariable`)?

```haskell
-- Remote meta-variables are meta-variables originating in other
-- modules. These meta-variables are always instantiated. We do not
-- retain all the information about a local meta-variable when
-- creating an interface:
--
-- * The 'mvPriority' field is not needed, because the meta-variable
--   cannot be instantiated.
-- * The 'mvFrozen' field is not needed, because there is no point in
--   freezing instantiated meta-variables.
-- * The 'mvListeners' field is not needed, because no meta-variable
--   should be listening to this one.
-- * The 'mvTwin' field is not needed, because the meta-variable has
--   already been instantiated.
-- * The 'mvPermutation' is currently removed, but could be retained
--   if it turns out to be useful for something.
-- * The only part of the 'mvInfo' field that is kept is the
--   'miModality' field. The 'miMetaOccursCheck' and 'miGeneralizable'
--   fields are omitted, because the meta-variable has already been
--   instantiated. The 'Range' that is part of the 'miClosRange' field
--   and the 'miNameSuggestion' field are omitted because instantiated
--   meta-variables are typically not presented to users. Finally the
--   'Closure' part of the 'miClosRange' field is omitted because it
--   can be large (at least if we ignore potential sharing).
```
This comment could quickly become out of date, but perhaps that is fine.

nad commented:
I added the comment above.

I also fixed the `size-solver` test case. However, I think we should consider removing this program from the repository, like we have with other programs.

nad commented:
I think this code is ready to be merged, except perhaps that there is only one (trivial) test case for `--save-metas`. All tests go through with `--save-metas` (except for one test case that includes the text `instantiating all meta variables` in the output), but I do not want to run all tests twice. Any suggestions?

nad commented:
One option would be to use `--save-metas` for the tests involving the standard library, to ensure that the tests involve lots of imported modules, as well as compilation. Does this sound reasonable?

nad commented:
> One option would be to use `--save-metas` for the tests involving the standard library, to ensure that the tests involve lots of imported modules, as well as compilation. Does this sound reasonable?

Maybe not, I think we should test the "default" way of using the standard library. I added the flag to some compiler test cases with imports.

andreasabel commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -3870,8 +3961,12 @@ enableCaching = optCaching <$> pragmaOptions
-  , redSt  :: TCState  -- ^ Read only access to state (signature, metas...).
+  { redEnv  :: TCEnv    -- ^ Read only access to environment.
+  , redSt   :: TCState  -- ^ Read only access to state (signature, metas...).
+  , redPred :: Maybe (MetaId -> ReduceM Bool)
```
This makes `ReduceM` a recursive type.
Pervers.
